file_catalog/server.py

```diff
@@ -474,43 +476,39 @@ async def post(self) -> None:
         if 'uuid' not in metadata:
             metadata['uuid'] = str(uuid1())
 
+        # Validate Incoming Data
         if not self.validation.validate_metadata_creation(self, metadata):
             return
 
         set_last_modification_date(metadata)
-        if await pathfinder.contains_existing_filepaths(self, metadata):
-            return
-
-        db_file = await self.db.get_file({'uuid': metadata['uuid']})
 
-        if db_file:
-            # file uuid already exists, check checksum
-            if db_file['checksum'] != metadata['checksum']:
-                # the uuid already exists (no replica since checksum is different
-                self.send_error(409, reason='Conflict with existing file (uuid already exists)',
-                                file=os.path.join(self.files_url, db_file['uuid']))
-                return
-            elif any(f in db_file['locations'] for f in metadata['locations']):
-                # replica has already been added
-                self.send_error(409, reason='Replica has already been added',
-                                file=os.path.join(self.files_url, db_file['uuid']))
+        # Deconflict with DB Records
+        # NOTE - POST should not conflict with any existing record
+        # NOTE - by uuid, by existing location(s), or by existing file-version
+        if await self.db.get_file({'uuid': metadata['uuid']}):
+            self.send_error(
+                409,
+                reason='Conflict with existing file (uuid already exists)',
+                file=os.path.join(self.files_url, metadata['uuid'])
+            )
+            return
+        if await deconfliction.any_location_in_db(self, metadata.get("locations")):
+            return
+        try:  # check if `metadata` will conflict with an existing metadata record
+            if await deconfliction.FileVersion(metadata).is_in_db(self):
                 return
-            else:
-                # add replica
-                db_file['locations'].extend(metadata['locations'])
+        except deconfliction.IndeterminateFileVersionError:
+            self.send_error(400, reason="File-version cannot be detected from the given 'metadata'")
 
-                await self.db.update_file(db_file['uuid'], {'locations': db_file['locations']})
-                self.set_status(200)
-                uuid = db_file['uuid']
-        else:
-            uuid = await self.db.create_file(metadata)
-            self.set_status(201)
+        # Create & Write-Back
+        await self.db.create_file(metadata)
+        self.set_status(201)
```

Comment:
I see we lost the "additional replica" of this POST route, but I'm pretty sure we talked about how this route would handle new records and `/locations` would handle adding locations.

Reply:
Can you elaborate on this?

Reply:
On old line 486, we start two paths:
- If we found a file by that UUID, we check to see if we're breaking something (409), or if we add a new 'locations' entry `# add replica` (200)

- If we didn't find a file by that UUID, we create a new record (201)


The new code checks to make sure the POST body is
- not providing anything that shouldn't be provided (i,e.: mongo_id, meta_modify_date)
- providing everything that should be provided (i.e.: logical_name, locations, checksum, etc.)
- not overwriting an existing uuid
- not matching an existing location
- not matching an existing file-version (logical_name, checksum)

Once it passes the gauntlet, the new metadata record is created (201).


Functionally, this POST route lost that `# add replica` (200):
- there is no way to get an additional location in an existing record and a 200 code out of this route now
 
However, I think we discussed that, and how LTA was using the `.../{uuid}/locations` route exclusively for adding new locations. So that add replica functionality wasn't used in this route and isn't necessary.

Not suggesting any code change here; I think it looks good. I just wanted to confirm that's the change we were talking about and make sure it's intentional (I think it is?)

Reply:
I confirm that's the intended change.